docs: Release 1.33 news doc (#168)
nginx/unit-docs: source/news/2024/unit-1.33.0-released.rst

:orphan:

####################
Unit 1.33.0 Released
####################

Unit 1.33.0 is now available. This release includes a number of new features
and changes:

************************************************************************
Feature 1
************************************************************************

Feature 1 description.

*******************************************************************
Feature 2
*******************************************************************

Feature 2 description.

*************************************
Changes in behavior and other updates
*************************************

==========================================================================
Change 1
==========================================================================

Change 1 description.

=======================================================
Change 2
=======================================================

Change 2 description.

************
Wall of fame
************

Special Thanks to all external contributors helping us

> making Unit better! With 1.32.0 we would like to send a shout out to:

Reviewer: 1.33.0

Author: ```suggestion
making Unit better! With 1.33.0 we would like to send a shout out to:
```

:orphan:

####################
Unit 1.33.0 Released
####################

We are pleased to announce the release of NGINX Unit 1.33.0. This release includes
a number of new features and changes:

*************************
New configuration options
*************************

> This release introduces two new configuration options:

Reviewer: s/two/three

#. **listen_threads**

   This option can be set under **/settings/listen_threads** and controls the
   number of threads the router process creates to handle client

>    connections. By default Unit creates the same number of threads as there

Reviewer 2: ```suggestion
   connections. By default, Unit creates the same number of threads as there
```

   are CPUs available.

#. **backlog**

   This option can be set under **/listeners/backlog**. This is a per-listener
   option that sets the the backlog parameter as passed to the **listen(2)**
   system-call, which defines the maximum length for the queue of pending
   connections for the socket.

   This is analogous to the **backlog** parameter of the **listen** directive in
   NGINX.

> 

Reviewer: Format as required...

```
 3) factory                                                                     
                                                                                
This can be set under '/applications/<app>/factory'                             
                                                                                
This is a boolean value and if set to 'true', Unit treats 'callable' as         
a factory.                                                                      
                                                                                
The default is 'false'.                                                         
                                                                                
NOTE: Unit does *not* support passing arguments to factories.
```

****************
unitctl CLI tool
****************

> :ref:`unitctl <unitctl>` is a powerful, Rust-based CLI tool that allows you to

Reviewer 2: what does "powerful" mean? it's markety sounding.

Author: This is in line with the tone and style used in these release blogs previously.

deploy, manage, and configure Unit in your environment.

****************************
Chunked request body support
****************************

Unit can now accept chunked requests rather than returning **411

> Length Required**. This feature is experimental and can

Reviewer 2: What does "experimental" mean? Is there any risk that needs to be called out?

Author: @ac000 might be able to address that

Reviewer: Here 'experimental' is the qualifier for why it's not enabled by default.

It's _experimental_ in that it was a quick _temporary_ implementation to fulfil a specific requirement, until a more permanent solution is made.

Author: why not say that instead?

```suggestion
Length Required**. This is a temporary workaround while we work on a long-term improvement, and can
```

Reviewer: Think I prefer it the way it is... the alternative text is going too far into the weeds for me...

Author: we need to explain what experimental implies one way or another

Reviewer: It's not documented and subject to change...

be enabled setting the **/settings/chunked_transform** configuration option
to true.

*************************************
Changes in behavior and other updates
*************************************

> * On Linux, we now default to a **listen(2)** backlog of -1, which means we
>   use the OS's default: 4096 for Linux 5.4 and later; 128 for older versions.

Reviewer 2: ```suggestion
* On Linux, we now default to a **listen(2)** backlog of `-1`, which means we
  use the OS's default: `4096` for Linux 5.4 and later; `128` for older versions.
```

* The Python language module now supports **Application Factories**
  (thanks to Gourav).

**********************
Changes for developers
**********************

We have made some changes to the build system:

===============================================
Prettified make output by default with GNU make
===============================================

Instead of getting the normal compiler command for each target being built
you now get a simplified line like:

.. code-block:: console

   CC     build/src/nxt_cgroup.o

You can use the `V=1` option to get the old verbose output, for example:

.. code-block:: console

   make V=1

==============
Make variables
==============

You can now control some aspects of the build process by passing variables to
make (like the above). The currently supported variables are:

.. list-table::
   :widths: 15 80 5
   :header-rows: 1

   * - Option
     - Description
     - Default
   * - **D=1**
     - Enables debug builds (-O0)
     - 0
   * - **E=0**
     - Disables -Werror
     - 1
   * - **V=1**
     - Enables verbose output
     - 0
   * - **EXTRA_CFLAGS=**
     - Add extra compiler options
     -

===========
GCC & Clang
===========

> We removed support for a bunch of esoteric compilers. GCC and Clang are now the
> only supported compilers for building Unit.

Reviewer 2: ```suggestion
We removed support for a several lesser-known compilers. GCC and Clang are now the
only supported compilers for building Unit.
```

Reviewer: > We removed support for a several lesser-known compilers. GCC and Clang are now the

At least remove the 'a'!

Reviewer 2: Oops -- thanks for catching that! Fixed in the edit.

==========
-std=gnu11
==========

We now build with **-std=gnu11** (C11 with GNU extensions). While previously we

> didn't explicitly set the -std= option, as we were supporting CentOS 7 (which is now

Reviewer 2: there's inconsisencies in how options are style. Sometimes, they're bold. Sometimes, they're in-line code. Sometimes, they're plain text. I recommend in-line code for all of them.

Above, it's in-line code.

<img width="581" alt="image" src="https://github.com/user-attachments/assets/d80c04c2-b83f-47d3-82ee-cb8e4fb7c7aa">

Author: I went with bold for consistency with the other release documents.

If there were still options using inline code, it's not intentional

EOL), we were effectively limited to **-std=gnu89/90**.

************
Wall of fame
************

Special Thanks to all external contributors helping us
making Unit better! With 1.33.0 we would like to send a shout out to:

- Alejandro Colomar
- Costas Drongos
- Gourav

> - Remi Collet

Reviewer: There will hopefully be another name to add to this list!

Reviewer: OK, you can add

```
Robbie McKinstry
```

to the list.

> Special thanks to Arjun for his fuzzing work.

Reviewer 2: Is "fuzzing" a generally known term for our global audience?

**************
Full Changelog
**************

.. code-block:: none

  Changes with Unit 1.33.0                                         29 Aug 2024

    *) Feature: show list of loaded language modules in the /status
       endpoint.

    *) Feature: make the number of router threads configurable.

    *) Feature: make the listen(2) backlog configurable.

    *) Feature: add fuzzing via oss-fuzz.

    *) Feature: add Python application factory support.

    *) Feature: add chunked request body support.

    *) Feature: add "if" option to the "match" object.

    *) Feature: Unit ships with a new Rust based CLI application "unitctl".

> 

Reviewer: Another feature to add

```
    *) Feature: the wasm-wasi-component language module now inherits the
       processes environment.
```

   are CPUs available.

#. **backlog**

   This option can be set under **/listeners/backlog**. This is a per-listener
   option that sets the the backlog parameter as passed to the **listen(2)**
   system-call, which defines the maximum length for the queue of pending
   connections for the socket.

   This is analogous to the **backlog** parameter of the **listen** directive in
   NGINX.

#. **factory**

  This can be set under '/applications/<app>/factory'

  This option is a boolean value. If set to 'true', Unit treats 'callable' as
  a factory.

  The default value is 'false'.

  NOTE: Unit does **not** support passing arguments to factories.

****************
unitctl CLI tool
****************

:ref:`unitctl <unitctl>` is a Rust-based CLI tool that allows you to
deploy, manage, and configure Unit in your environment.

****************************
Chunked request body support
****************************

Unit can now accept chunked requests rather than returning **411
Length Required**. This feature is experimental (not documented and subject to change), and can
be enabled setting the **/settings/chunked_transform** configuration option

> to true.

Reviewer: 'true' should probably be bold.

Author: ```suggestion
to **true**.
```

       endpoint.

    *) Feature: make the number of router threads configurable.

    *) Feature: make the listen(2) backlog configurable.

    *) Feature: add fuzzing via oss-fuzz.

    *) Feature: add Python application factory support.

    *) Feature: add chunked request body support.

    *) Feature: add "if" option to the "match" object.

    *) Feature: Unit ships with a new Rust based CLI application "unitctl".

    *) Feature: the wasm-wasi-component language module now inherits the
       processes environment.

    *) Change: under systemd unit runs in forking mode (once again).

    *) Change: if building with njs, version 0.8.3 or later is now required.

    *) Change: Unit now builds with -std=gnu11 (C11 with GNU extensions).

    *) Change: Unit now creates the full directory path for the PID file and
       control socket.

    *) Change: build system improvements, including pretty printing the make
       output and enabling various make variables to influence the build
       process (see: make help).

    *) Change: better detection of available runnable CPUs on Linux.

    *) Change: default listen(2) backlog on Linux now defaults to Kernel
       default.

> 

Reviewer: New bugfix

```
*) Bugfix: fix a crash when interrupting a download via a proxy.
```
